odoo/fields.py

```diff
@@ -1409,6 +1410,14 @@ def convert_to_column(self, value, record, values=None, validate=True):
             # cache, and would be overridden by the value read from database!
             currency = record[:1].with_context(prefetch_fields=False)[self.currency_field]
 
+        if not currency and currency_field.related and values and currency_field.related[0] in values:
+            related_field_name = currency_field.related[0]
+            related_field = record._fields[related_field_name]
+            related_model_name = related_field.comodel_name
+            currency = record.env[related_model_name].browse(values[related_field_name])
```

Comment:
Cannot we use the the convert_to_cache + convert_to_record to do it ? it avoids to reproduce the same behavior than the convert_to_record of X2one relation.